### ISSUE TITLE
[Bug] register_full_backward_hook modifies module output type

### ISSUE THREAD

**Issue by reinaw1012** [pytorch/pytorch#61026]
## 🐛 Bug

register_full_backward_hook forces the output of the module it's applied on to be of type `torch.Tensor`, even if the module is defined so that the proper output should be a subclass of `torch.Tensor`. 

In the code sample below, the model `SimpleModel()` should return a `DummyTensor`, but after the backward hook is applied, the return type is `torch.Tensor`.

## To Reproduce

```python
import torch
import torch.nn as nn
        
class DummyTensor(torch.Tensor):
    def __init__(self, *args, **kwargs):
        super().__init__()
        
class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(2, 2)
    def forward(self, x):
        return DummyTensor(self.fc(x))
    
data = torch.Tensor([[1, 2],[3, 4],[5, 6]])
model = SimpleModel()
before_hook_output = model(data)

def backward_hook(module, grad_input, grad_output):
    return #do nothing
model.register_full_backward_hook(backward_hook)

after_hook_output = model(data)

print("Model output type before applying wrapper: ", type(before_hook_output))
print("Model output type after applying wrapper: ", type(after_hook_output))
```

Outputs:
```
Model output type before applying wrapper:  <class '__main__.DummyTensor'>
Model output type after applying wrapper:  <class 'torch.Tensor'>
```

## Expected Behavior
The output should be:
```
Model output type before applying wrapper:  <class '__main__.DummyTensor'>
Model output type after applying wrapper:  <class '__main__.DummyTensor'>
```

## Environment
```
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.6.12 |Anaconda, Inc.| (default, Sep  9 2020, 00:29:25) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19041-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.9.0
[pip3] torchaudio==0.9.0
[pip3] torchvision==0.10.0
```

cc @AdlyTempleton

**Comment by VitalyFedyunin** [comment]
I was trying to find a better label, but failed to do so, feel free to return it back to the triage pool.

**Comment by albanD** [comment]
As of 2.1, this has been fixed. Most likely because we changed the default behavior of torch_function subclass to preserve itself.
